src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -63,16 +62,76 @@ case class ArrayIteratorTriplet(calcLength: Code[Unit], length: Option[Code[Int]
 }
 
 private class Emit(
-  fb: FunctionBuilder[_],
-  mb: StagedBitSet,
+  mb: MethodBuilder,
   tAggInOpt: Option[TAggregable],
   nSpecialArguments: Int) {
 
+  private val maxBytecodeSizeTarget: Int = 2048
+  private val opSize: Int = 20
+
+  def chunkIRs(irs: Seq[IR]): Array[Seq[IR]] = {
+    val chunkBounds = getChunkBounds(irs.map(_.size * opSize))
+    println(chunkBounds.mkString(","))
+    Array.tabulate(chunkBounds.length - 1)(i => irs.slice(chunkBounds(i), chunkBounds(i + 1)))
+  }
+
+  def getChunkBounds(sizes: Seq[Int]): Array[Int] = {
+    var total = 0
+    val ab = new ArrayBuilder[Int]()
+    ab += 0
+    sizes.zipWithIndex.foreach { case (size, i) =>
+      if (total + size <= maxBytecodeSizeTarget)
+        total += size
+      else {
+        if (total == 0)
+          total += size
+        else
+          ab += i
+        total = 0
+      }
+    }
+    ab += sizes.size
+    ab.result()
+  }
+
   val methods: mutable.Map[String, Seq[(Seq[Type], MethodBuilder)]] = mutable.Map().withDefaultValue(Seq())
 
   import Emit.E
   import Emit.F
 
+  private def wrapToMethod(irs: Seq[IR], env: E)(useValues: (MethodBuilder, Type, EmitTriplet) => Code[Unit]): Code[Unit] = {
+    val newMB = mb.fb.newMethod(mb.parameterTypeInfo, typeInfo[Unit])
+    val emitWrapper = new Emit(newMB, tAggInOpt, nSpecialArguments)
+    val c =
+      for (ir <- irs) yield {
+        val v = emitWrapper.emit(ir, env)
+        useValues(newMB, ir.typ, v)
+      }
+    newMB.emit(Code(c: _*))
+    val args = mb.parameterTypeInfo.zipWithIndex.map { case (ti, i) => mb.getArg(i + 1)(ti).load() }
+    newMB.invoke(args: _*)
+  }
+
+  private def treeWrapToMethod(irs: Seq[IR], env: E)(useValues: (MethodBuilder, Type, EmitTriplet) => Code[Unit]): Code[Unit] = {
```

Comment:
Squinting a bit, it looks like you might be able to unify these to and make it a tad simpler (but maybe not, you're chunking two different things, IRs and invokes.)